packages/astro/src/core/build/plugins/plugin-content.ts

```diff
@@ -0,0 +1,302 @@
+			]);
+			newManifest.serverEntries = Array.from(serverComponents);
+			newManifest.clientEntries = Array.from(clientComponents);
+			await fsMod.promises.mkdir(contentCacheDir, { recursive: true });
+			await fsMod.promises.writeFile(contentManifestFile, JSON.stringify(newManifest), { encoding: 'utf8' });
+
+			const cacheExists = fsMod.existsSync(cache);
+			fsMod.mkdirSync(cache, { recursive: true })
+			await fsMod.promises.mkdir(cacheTmp, { recursive: true });
+			await copyFiles(distContentRoot, cacheTmp, true);
+			if (cacheExists) {
+				await copyFiles(contentCacheDir, distContentRoot, false);
+			}
+			await copyFiles(cacheTmp, contentCacheDir);
+			await fsMod.promises.rm(cacheTmp, { recursive: true, force: true });
+		}
+	};
+}
+
+const entryCache = new Map<string, string>();
+function findEntryFromSrcRelativePath(lookupMap: ContentLookupMap, srcRelativePath: string) {
+	let value = entryCache.get(srcRelativePath);
+	if (value) return value;
+	for (const collection of Object.values(lookupMap)) {
+		for (const entry of Object.values(collection)) {
+			for (const entryFile of Object.values(entry)) {
+				if (entryFile === srcRelativePath) {
+					value = entryFile;
+					entryCache.set(srcRelativePath, entryFile);
+					return value;
+				}
+			}
+		}
+	}
+}
+
+interface ContentEntries {
+	restoreFromCache: ContentManifestKey[];
+	buildFromSource: ContentManifestKey[];
+}
+function getEntriesFromManifests(oldManifest: ContentManifest, newManifest: ContentManifest): ContentEntries {
+	const { version: oldVersion, entries: oldEntries } = oldManifest;
+	const { version: newVersion, entries: newEntries } = newManifest;
+	let entries: ContentEntries = { restoreFromCache: [], buildFromSource: [] };
+
+	const newEntryMap = new Map<ContentManifestKey, string>(newEntries);
+	if (oldVersion !== newVersion || oldEntries.length === 0) {
+		entries.buildFromSource = Array.from(newEntryMap.keys());
+		return entries;
+	}
+	const oldEntryHashMap = new Map<string, ContentManifestKey>(oldEntries.map(([key, hash]) => [hash, key]))
+
+	for (const [entry, hash] of newEntryMap) {
+		if (oldEntryHashMap.has(hash)) {
+			entries.restoreFromCache.push(entry);
+		} else {
+			entries.buildFromSource.push(entry);
+		}
+	}
+	return entries;
+}
+
+async function generateContentManifest(opts: StaticBuildOptions, lookupMap: ContentLookupMap): Promise<ContentManifest> {
+	let manifest: ContentManifest = { version: CONTENT_MANIFEST_VERSION, entries: [], serverEntries: [], clientEntries: [] };
+	const limit = pLimit(10);
+	const promises: Promise<void>[] = [];
+
+	for (const [collection, { type, entries }] of Object.entries(lookupMap)) {
+		for (const entry of Object.values(entries)) {
+			const key: ContentManifestKey = { collection, type, entry };
+			const fileURL = new URL(encodeURI(joinPaths(opts.settings.config.root.toString(), entry)));
+			promises.push(limit(async () => {
+				const data = await fsMod.promises.readFile(fileURL, { encoding: 'utf8' });
```

Comment:
Is this operation safe? Should we check if `fileURL` exists and if we can read the file?